### PR TITLE
Cast model_path to string

### PR DESCRIPTION
**Pull request by Henrike100**
**Proposed changes**:
- Change `Agent.load` method to support `pathlib` paths, resolving issue #4341

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by sara-tagger** [comment]
Thanks for submitting a pull request 🚀 @JustinaPetr will take a look at it as soon as possible ✨

File: rasa/core/agent.py [hunk 1]
```diff
@@ -408,7 +408,7 @@ def load(
             if not os.path.exists(model_path):
                 raise ModelNotFound(f"No file or directory at '{model_path}'.")
             if os.path.isfile(model_path):
-                model_path = get_model(model_path)
+                model_path = get_model(str(model_path))
```
Comment (ricwo) [inline rasa/core/agent.py:411]: The type annotation for the `model_path` argument needs to be updated

**Review by ricwo — CHANGES_REQUESTED** [review]
Thanks a lot for submitting this PR 👍

File: rasa/core/agent.py (hunk 1 above)
Comment (silvasara) [reply rasa/core/agent.py:411]: We have checked the signatures of the related functions (```load()``` and ```get_model()```) their type annotations seem to be correct, so we didn't understand the suggested modification. Could you please clarify?

Comment (ricwo) [reply rasa/core/agent.py:411]: sure - casting the model path to `str` would only be necessary if `model_path` is of type `pathlib.Path`. the type annotations in the `load` function only have `model_path: Text`

Comment (Henrike100) [reply rasa/core/agent.py]: We added type annotation to `pathlib.Path` and added type checking. When you can, check if the changes are correct, please.

File: rasa/core/agent.py [hunk 2]
```diff
@@ -408,7 +409,10 @@ def load(
             if not os.path.exists(model_path):
```
Comment (ricwo) [inline rasa/core/agent.py:409]:
```suggestion
            model_path = Path(model_path)
            if not model_path.exists():
```

File: rasa/core/agent.py [hunk 3]
```diff
@@ -408,7 +409,10 @@ def load(
...
                 raise ModelNotFound(f"No file or directory at '{model_path}'.")
             if os.path.isfile(model_path):
```
Comment (ricwo) [inline rasa/core/agent.py:411]:
```suggestion
            if model_path.is_file():
```

File: rasa/core/agent.py [hunk 4]
```diff
@@ -408,7 +409,10 @@ def load(
...
-                model_path = get_model(model_path)
+                if isinstance(model_path, pathlib.Path):
+                    model_path = get_model(str(model_path))
+                else:
+                    model_path = get_model(model_path)
```
Comment (ricwo) [inline rasa/core/agent.py:415]:
```suggestion
                model_path = get_model(model_path)
```

Comment (ricwo) [reply rasa/core/agent.py:415]: `get_model()` should receive similar support for `Path` objects

File: rasa/core/agent.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import os
 import shutil
 import tempfile
+import pathlib
```
Comment (ricwo) [inline rasa/core/agent.py:6]:
```suggestion
from pathlib import Path
```

File: rasa/core/agent.py [hunk 6]
```diff
@@ -391,7 +392,7 @@ def update_model(
     @classmethod
     def load(
         cls,
-        model_path: Text,
+        model_path: Union[Text, pathlib.Path],
```
Comment (ricwo) [inline rasa/core/agent.py:395]:
```suggestion
        model_path: Union[Text, Path],
```

**Review by ricwo — CHANGES_REQUESTED** [review]
Thanks for making those changes 👍 

It would be great if you could add `Path` support to `get_model()` as well, so we don't do unnecessary conversions between the two functions.

Where did you come across having to use `Agent.load()` with a `Path` object? Did you use it somewhere in your own code, or is this an error you saw in the rasa codebase?

**Comment by Henrike100** [comment]
Thanks for the review, we will make the changes soon.

> Where did you come across having to use Agent.load() with a Path object?

We don't. The owner of the issue said:

> Adding support for pathlib paths would be quite convenient.

**Comment by silvasara** [comment]
> 
> It would be great if you could add `Path` support to `get_model()` as well, so we don't do unnecessary conversions between the two functions.
> 

For these suggestions we would have to change the `get_model` so that it accepts both Text and Path, but we already do this treatment in the `load`, so we don't understand the need for the change, could you elaborate better?

**Comment by ricwo** [comment]
@Henrike100 can you say more about why this would be convenient? are you calling `Agent.load` from your own code?

@silvasara Updating `get_model()` doesn't necessarily have to happen in this PR. The suggestions I made are still possible if you include a `str(model_path)` conversion before passing the parameter into `get_model`

**Comment by Henrike100** [comment]
> @Henrike100 can you say more about why this would be convenient? are you calling `Agent.load` from your own code?

I'm just solving issue #4341. It was the owner of the issue who said:

> Adding support for pathlib paths would be quite convenient.

I don't know why it would be convenient. It was the owner of the issue who said it would be. I haven't used `Agent.load` anywhere.

> The suggestions I made are still possible if you include a `str(model_path)` conversion before passing the parameter into `get_model`

Line 413 already does this:

`model_path = get_model(str(model_path))`

I can do this in a different way if you want. I can also make the other suggestions.

**Comment by ricwo** [comment]
> I'm just solving issue #4341. It was the owner of the issue who said:

It would be great if you could add a changelog entry referencing this issue

> Line 413 already does this:

yes but the prior `if` is not necessary - you can just pass `str(model_path)` into `get_model()`, regardless of whether its type is Text or Path

**Comment by Henrike100** [comment]
I made the changes you suggested, except for those involving `get_model`.

> It would be great if you could add a changelog entry referencing this issue

I already wrote the changelog. The file name is: `4341.improvement.md`.
Are you talking about this file?

**Review by ricwo — APPROVED** [review]
looks great, and thanks for your contribution 🎉

**Comment by ricwo** [comment]
@Henrike100 would you mind merging in master once again?

**Comment by Henrike100** [comment]
@ricwo Merged.

**Pull request merged**